File: src/main/java/com/uiptv/util/UiptUtils.java
package com.uiptv.util;

import com.uiptv.service.AccountService;

import java.net.URI;
import java.net.URL;
import java.util.HashMap;
import java.util.Map;
import java.util.regex.Pattern;

import static com.uiptv.util.StringUtils.EMPTY;
import static com.uiptv.util.StringUtils.isBlank;

public class UiptUtils {
    private static final String MAC_ADDRESS_REGEX = "^([0-9A-Fa-f]{2}[:-])"
            + "{5}([0-9A-Fa-f]{2})|"
            + "([0-9a-fA-F]{4}\\."
            + "[0-9a-fA-F]{4}\\."
            + "[0-9a-fA-F]{4})$";

    public static final String SPACER = " ";

    public static String replaceAllNonPrintableChars(String line) {
        try {
            return isBlank(line) ? EMPTY : line.replaceAll("\\p{Cntrl}", SPACER).replaceAll("[^\\p{Print}]", SPACER).replaceAll("\\p{C}", SPACER);
        } catch (Exception ignored) {
        }
        return line;
    }

    public static boolean isValidURL(String urlString) {
        try {
            URL url = new URL(urlString);
            url.toURI();
            return true;
        } catch (Exception ignored) {
        }
        return false;
    }

    public static boolean isValidMACAddress(String line) {
        if (isBlank(line)) return false;
        return Pattern.compile(MAC_ADDRESS_REGEX).matcher(line).matches();
    }

    public static String getUniqueNameFromUrl(String urlString) {
        try {
            URL url = new URL(urlString);
            URI uri = url.toURI();
            int i = 1;
            String validName;
            do {
                validName = uri.getHost() + " (" + i++ + ")";
            } while (AccountService.getInstance().getByName(validName) != null);
            return validName;
        } catch (Exception ignored) {
        }
        return urlString;
    }

    public static String getNameFromUrl(String urlString) {
        try {
            return new URL(urlString).toURI().getHost();
        } catch (Exception ignored) {
        }
        return urlString;
    }

    public static boolean isUrlValidXtremeLink(String urlString) {
        try {
            String queryString = new URL(urlString).toURI().getQuery();
            if (isBlank(queryString)) return false;
            return urlString.toLowerCase().contains("get.php?") && queryString.toLowerCase().contains("username") && queryString.toLowerCase().contains("password");
        } catch (Exception ignored) {
        }
        return false;
    }

    public static String getPathFromUrl(String urlString) {
        try {
            return urlString.split("get.php?")[0];
        } catch (Exception ignored) {
        }
        return urlString;
    }

    public static String getUserNameFromUrl(String urlString) {
        try {
            String queryString = new URL(urlString).toURI().getQuery();
            return getQueryMap(queryString).get("username");
        } catch (Exception ignored) {
        }
        return urlString;
    }

    public static String getPasswordNameFromUrl(String urlString) {
        try {
            String queryString = new URL(urlString).toURI().getQuery();
            return getQueryMap(queryString).get("password");
        } catch (Exception ignored) {
        }
        return urlString;
    }

    private static Map<String, String> getQueryMap(String query) {
        String[] params = query.split("&");
        Map<String, String> map = new HashMap<String, String>();
        for (String param : params) {
            String[] p = param.split("=");
            String name = p[0];
            if (p.length > 1) {
                String value = p[1];
                map.put(name, value);
            }
        }
        return map;
    }

    public static String sanitizeStalkerText(String text) {
        if (isBlank(text)) return EMPTY;
        StringBuilder sb = new StringBuilder();
        text.codePoints().forEach(cp -> {
            sb.appendCodePoint(mapSpecialCodePoint(cp));
        });
        return sb.toString();
    }

    private static int mapSpecialCodePoint(int c) {
        // Mathematical Bold Capital
        if (c >= 0x1D400 && c <= 0x1D419) return 'A' + (c - 0x1D400);
        // Mathematical Bold Small
        if (c >= 0x1D41A && c <= 0x1D433) return 'a' + (c - 0x1D41A);
        // Mathematical Bold Digits
        if (c >= 0x1D7CE && c <= 0x1D7D7) return '0' + (c - 0x1D7CE);
        // Negative Circled Capital
        if (c >= 0x1F150 && c <= 0x1F169) return 'A' + (c - 0x1F150);
        // Negative Squared Capital
        if (c >= 0x1F170 && c <= 0x1F189) return 'A' + (c - 0x1F170);

        switch (c) {
            case 0x029C: return 'H'; // ʜ
            case 0x1D0F: return 'O'; // ᴏ
            case 0x1D1B: return 'T'; // ᴛ
            case 0x1D07: return 'E'; // ᴇ
            case 0x1D00: return 'A'; // ᴀ
            case 0x029F: return 'L'; // ʟ
            case 0x1D18: return 'P'; // ᴘ
            case 0x0280: return 'R'; // ʀ
            case 0x1D1C: return 'U'; // ᴜ
            case 0x0274: return 'N'; // ɴ
            case 0x1D0D: return 'M'; // ᴍ
            case 0x026A: return 'I'; // ɪ
            case 0x1D04: return 'C'; // ᴄ
            case 0x1D20: return 'V'; // ᴠ
            case 0x1D05: return 'D'; // ᴅ
            case 0x2776: return '1'; // ❶
            case 0x2777: return '2'; // ❷
            case 0x278C: return '3'; // ➌
            case 0x279F: return ' '; // ➩
            case 0x27A4: return ' '; // ➤
            case 0x27D0: return ' '; // ⟐
            case 0x1F511: return ' '; // 🔑
            case 0x1F194: return ' '; // 🆔
            case 0x1F4DD: return ' '; // 📝
            case 0x1F538: return ' '; // 🔸
            case 0x25CF: return ' '; // ●
            case 0x251C: return ' '; // ├
            case 0x2500: return ' '; // ─
            case 0x2502: return ' '; // │
            case 0x2570: return ' '; // ╰
            case 0x256D: return ' '; // ╭
            case 0x1F3C1: return ' '; // 🏁
            case 0x23F0: return ' '; // ⏰
            case 0x1F47D: return ' '; // 👽
            case 0x1F510: return ' '; // 🔐
            case 0x1FA62: return ' '; // 🪢
            case 0x1F3AF: return ' '; // 🎯
            case 0x1F4EE: return ' '; // 📮
            case 0x1F310: return ' '; // 🌐
            case 0x26D1: return ' '; // ⛑
            case 0x1F30D: return ' '; // 🌍
            case 0x269C: return ' '; // ⚜
            case 0x2620: return ' '; // ☠
            case 0x2605: return ' '; // ★
            case 0x2606: return ' '; // ☆
            case 0x1F5A5: return ' '; // 🖥
            default: return c;
        }
    }
}

File: server/src/main/kotlin/com/uiptv/util/UiptUtils.kt
package com.uiptv.util

import com.uiptv.service.AccountService
import java.net.URI
import java.util.regex.Pattern

object UiptUtils {
    private const val MAC_ADDRESS_REGEX = "^(?:([0-9A-Fa-f]{2}[:-]){5}([0-9A-Fa-f]{2})|([0-9a-fA-F]{4}\\.[0-9a-fA-F]{4}\\.[0-9a-fA-F]{4}))$"
    const val SPACER = " "
    private val ILLEGAL_URI_CHAR_REPLACEMENTS = mapOf(
        " " to "%20",
        "|" to "%7C",
        "<" to "%3C",
        ">" to "%3E",
        "\"" to "%22",
        "{" to "%7B",
        "}" to "%7D",
        "\\" to "%5C",
        "^" to "%5E",
        "`" to "%60"
    )

    @JvmStatic
    fun replaceAllNonPrintableChars(line: String?): String? {
        return try {
            if (StringUtils.isBlank(line)) StringUtils.EMPTY else line!!.replace("\\p{Cntrl}".toRegex(), SPACER).replace("[^\\p{Print}]".toRegex(), SPACER).replace("\\p{C}".toRegex(), SPACER)
        } catch (_: Exception) {
            line
        }
    }

    @JvmStatic
    fun isValidURL(urlString: String?): Boolean {
        return try {
            val uri = parseUrlLikeUri(urlString)
            !StringUtils.isBlank(uri.scheme) && !StringUtils.isBlank(uri.host)
        } catch (_: Exception) {
            false
        }
    }

    @JvmStatic
    fun isValidMACAddress(line: String?): Boolean =
        !StringUtils.isBlank(line) && Pattern.compile(MAC_ADDRESS_REGEX).matcher(line).matches()

    @JvmStatic
    fun getUniqueNameFromUrl(urlString: String?): String {
        return try {
            val uri = parseUrlLikeUri(urlString)
            var index = 1
            var validName: String
            do {
                validName = uri.host + " (" + index++ + ")"
            } while (AccountService.getInstance().getByName(validName) != null)
            validName
        } catch (_: Exception) {
            urlString.orEmpty()
        }
    }

    @JvmStatic
    fun getNameFromUrl(urlString: String?): String {
        return try {
            parseUrlLikeUri(urlString).host
        } catch (_: Exception) {
            urlString.orEmpty()
        }
    }

    @JvmStatic
    fun isUrlValidXtremeLink(urlString: String?): Boolean {
        return try {
            val queryString = parseUrlLikeUri(urlString).query
            !StringUtils.isBlank(queryString) &&
                urlString!!.lowercase().contains("get.php?") &&
                queryString.lowercase().contains("username") &&
                queryString.lowercase().contains("password")
        } catch (_: Exception) {
            false
        }
    }

    @JvmStatic
    fun getPathFromUrl(urlString: String?): String {
        return try {
            urlString!!.split("get.php?")[0]
        } catch (_: Exception) {
            urlString.orEmpty()
        }
    }

    @JvmStatic
    fun getUserNameFromUrl(urlString: String?): String {
        return try {
            val queryString = parseUrlLikeUri(urlString).query
            getQueryMap(queryString)["username"].orEmpty()
        } catch (_: Exception) {
            urlString.orEmpty()
        }
    }

    @JvmStatic
    fun getPasswordNameFromUrl(urlString: String?): String {
        return try {
            val queryString = parseUrlLikeUri(urlString).query
            getQueryMap(queryString)["password"].orEmpty()
        } catch (_: Exception) {
            urlString.orEmpty()
        }
    }

    @JvmStatic
    fun parseUrlLikeUri(urlString: String?): URI {
        val normalized = urlString?.trim().orEmpty()
        require(normalized.isNotEmpty()) { "urlString cannot be blank" }
        return try {
            URI(normalized)
        } catch (_: Exception) {
            URI(sanitizeIllegalUriChars(normalized))
        }
    }

    private fun sanitizeIllegalUriChars(value: String): String {
        var sanitized = value
        ILLEGAL_URI_CHAR_REPLACEMENTS.forEach { (raw, escaped) ->
            sanitized = sanitized.replace(raw, escaped)
        }
        return sanitized
    }

    private fun getQueryMap(query: String): Map<String, String> {
        val params = query.split("&")
        val map = HashMap<String, String>()
        for (param in params) {
            val parts = param.split("=")
            val name = parts[0]
            if (parts.size > 1) {
                map[name] = parts[1]
            }
        }
        return map
    }

    @JvmStatic
    fun sanitizeStalkerText(text: String?): String {
        if (StringUtils.isBlank(text)) return StringUtils.EMPTY
        val sb = StringBuilder()
        text!!.codePoints().forEach { cp -> sb.appendCodePoint(mapSpecialCodePoint(cp)) }
        return sb.toString()
    }

    private fun mapSpecialCodePoint(c: Int): Int {
        if (c in 0x1D400..0x1D419) return 'A'.code + (c - 0x1D400)
        if (c in 0x1D41A..0x1D433) return 'a'.code + (c - 0x1D41A)
        if (c in 0x1D7CE..0x1D7D7) return '0'.code + (c - 0x1D7CE)
        if (c in 0x1F150..0x1F169) return 'A'.code + (c - 0x1F150)
        if (c in 0x1F170..0x1F189) return 'A'.code + (c - 0x1F170)

        return when (c) {
            0x029C -> 'H'.code
            0x1D0F -> 'O'.code
            0x1D1B -> 'T'.code
            0x1D07 -> 'E'.code
            0x1D00 -> 'A'.code
            0x029F -> 'L'.code
            0x1D18 -> 'P'.code
            0x0280 -> 'R'.code
            0x1D1C -> 'U'.code
            0x0274 -> 'N'.code
            0x1D0D -> 'M'.code
            0x026A -> 'I'.code
            0x1D04 -> 'C'.code
            0x1D20 -> 'V'.code
            0x1D05 -> 'D'.code
            0x2776 -> '1'.code
            0x2777 -> '2'.code
            0x278C -> '3'.code
            0x279F, 0x27A4, 0x27D0, 0x1F511, 0x1F194, 0x1F4DD, 0x1F538, 0x25CF,
            0x251C, 0x2500, 0x2502, 0x2570, 0x256D, 0x1F3C1, 0x23F0, 0x1F47D,
            0x1F510, 0x1FA62, 0x1F3AF, 0x1F4EE, 0x1F310, 0x26D1, 0x1F30D, 0x269C,
            0x2620, 0x2605, 0x2606, 0x1F5A5 -> ' '.code
            else -> c
        }
    }
}
